`unmanaged-cluster` merging kind configs

jpmcb opened the pull request:
## What this PR does / why we need it
Merges kind configs given via flags and the `ProviderConfiguration.rawKindConfig`

Merge order:
- Prefers "global" configs from the cluster config which is built off
  the CLI and env var configuration values 
- Configs given via the `ProviderConfiguration` field
- Default values

Merging nodes happens in a 1:1 order based on the index. Extra node configurations are ignored if there is a 1 to 1 mismatch (user does `--worker-node-count 2` but has 5 nodes in their `rawKindConfig`. The 3 extra nodes are ignored.)

## Which issue(s) this PR fixes
Fixes: #3755

## Describe testing done for PR

Given the following config, note that we are _not_ configuring the node image in the kind config:
```yaml
ClusterName: kind-test
KubeconfigPath: ""
ExistingClusterKubeconfig: ""
NodeImage: ""
Provider: kind
ProviderConfiguration:
  rawKindConfig: |
    nodes:
    - role: control-plane
      extraPortMappings:
      - containerPort: 888
        hostPort: 888
        listenAddress: "127.0.0.1"
        protocol: TCP
    - role: worker
    - role: worker
Cni: calico
CniConfiguration: {}
PodCidr: 10.244.0.0/16
ServiceCidr: 10.96.0.0/16
TkrLocation: ""
AdditionalPackageRepos: []
PortsToForward: []
SkipPreflight: false
ControlPlaneNodeCount: "1"
WorkerNodeCount: "0"
Profiles: []
```

Using this during bootstrapping:
```
❯ tanzu unmanaged-cluster create -f kind-test.yaml

📁 Created cluster directory
   Reading ProviderConfiguration from config file. Some provider specific flags and configs may be ignored.

🧲 Resolving and checking Tanzu Kubernetes release (TKr) compatibility file
   projects.registry.vmware.com/tce/compatibility
   Compatibility file exists at /home/jmcb/.config/tanzu/tkg/unmanaged/compatibility/projects.registry.vmware.com_tce_compatibility_v4

🔧 Resolving TKr
   projects.registry.vmware.com/tce/tkr:v0.17.0-dev-2
   TKr exists at /home/jmcb/.config/tanzu/tkg/unmanaged/bom/projects.registry.vmware.com_tce_tkr_v0.17.0-dev-2
   Rendered Config: /home/jmcb/.config/tanzu/tkg/unmanaged/kind-test/config.yaml
   Bootstrap Logs: /home/jmcb/.config/tanzu/tkg/unmanaged/kind-test/bootstrap.log

🔧 Processing Tanzu Kubernetes Release

🎨 Selected base image
   projects.registry.vmware.com/tce/kind:v1.22.4

📦 Selected core package repository
   projects.registry.vmware.com/tce/repo-10:0.10.0

📦 Selected additional package repositories
   projects.registry.vmware.com/tce/main:0.11.0

📦 Selected kapp-controller image bundle
   projects.registry.vmware.com/tce/kapp-controller-multi-pkg:v0.30.1

🚀 Creating cluster kind-test
   Cluster creation using kind!
   ❤️  Checkout this awesome project at https://kind.sigs.k8s.io
   Base image downloaded
   Cluster created
   To troubleshoot, use:

   kubectl ${COMMAND} --kubeconfig /home/jmcb/.config/tanzu/tkg/unmanaged/kind-test/kube.conf

📧 Installing kapp-controller
   kapp-controller status: Running

📧 Installing package repositories
   tkg-core-repository package repo status: Reconcile succeeded

🌐 Installing CNI
   calico.community.tanzu.vmware.com:3.22.1

✅ Cluster created

🎮 kubectl context set to kind-test

View available packages:
   tanzu package available list
View running pods:
   kubectl get po -A
Delete this cluster:
   tanzu unmanaged delete kind-test
```

We can see that it deployed the nodes with the default node image:
```
❯ k get nodes
NAME                      STATUS   ROLES                  AGE   VERSION
NAME                      STATUS     ROLES                  AGE   VERSION
kind-test-control-plane   NotReady   control-plane,master   88s   v1.22.4
kind-test-worker          NotReady   <none>                 69s   v1.22.4
kind-test-worker2         NotReady   <none>                 68s   v1.22.4

❯ docker ps
CONTAINER ID   IMAGE                                           COMMAND                  CREATED         STATUS         PORTS                                               NAMES
eff89fa20967   projects.registry.vmware.com/tce/kind:v1.22.4   "/usr/local/bin/entr…"   2 minutes ago   Up 2 minutes                                                       kind-test-worker
149ffab07b25   projects.registry.vmware.com/tce/kind:v1.22.4   "/usr/local/bin/entr…"   2 minutes ago   Up 2 minutes                                                       kind-test-worker2
dcd23bcd1ef0   projects.registry.vmware.com/tce/kind:v1.22.4   "/usr/local/bin/entr…"   2 minutes ago   Up 2 minutes   127.0.0.1:888->888/tcp, 127.0.0.1:36393->6443/tcp   kind-test-control-plane
```

If we inspect the file found in the cluster's config directory, we can can see what the kind config looked like:

```
❯ cat ~/.config/tanzu/tkg/unmanaged/kind-test/kindconfig.yaml
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
name: kind-test
nodes:
    - role: control-plane
      image: projects.registry.vmware.com/tce/kind:v1.22.4
      extraPortMappings:
        - containerPort: 888
          hostPort: 888
          listenAddress: 127.0.0.1
          protocol: TCP
    - role: worker
      image: projects.registry.vmware.com/tce/kind:v1.22.4
    - role: worker
      image: projects.registry.vmware.com/tce/kind:v1.22.4
networking:
    ipFamily: ipv4
    apiServerAddress: 127.0.0.1
    podSubnet: 10.244.0.0/16
    serviceSubnet: 10.96.0.0/16
    disableDefaultCNI: true
    kubeProxyMode: iptables
```

Using this configuration file, we can provide an extra port mapping that gets merged with the provider's kind config:
```
❯ tanzu unmanaged-cluster create -f kind-test.yaml -p 999:999
...

❯ docker ps
CONTAINER ID   IMAGE                                           COMMAND                  CREATED              STATUS              PORTS                                                                     NAMES
2bc495e0bf76   projects.registry.vmware.com/tce/kind:v1.22.4   "/usr/local/bin/entr…"   About a minute ago   Up About a minute                                                                             kind-test-worker
8f21fea53235   projects.registry.vmware.com/tce/kind:v1.22.4   "/usr/local/bin/entr…"   About a minute ago   Up About a minute                                                                             kind-test-worker2
1f7fb37b19c9   projects.registry.vmware.com/tce/kind:v1.22.4   "/usr/local/bin/entr…"   About a minute ago   Up About a minute   127.0.0.1:888->888/tcp, 0.0.0.0:999->999/tcp, 127.0.0.1:37011->6443/tcp   kind-test-control-plane
```

Config looks like:
```
❯ cat ~/.config/tanzu/tkg/unmanaged/kind-test/kindconfig.yaml
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
name: kind-test
nodes:
    - role: control-plane
      image: projects.registry.vmware.com/tce/kind:v1.22.4
      extraPortMappings:
        - containerPort: 888
          hostPort: 888
          listenAddress: 127.0.0.1
          protocol: TCP
        - containerPort: 999
          hostPort: 999
    - role: worker
      image: projects.registry.vmware.com/tce/kind:v1.22.4
    - role: worker
      image: projects.registry.vmware.com/tce/kind:v1.22.4
networking:
    ipFamily: ipv4
    apiServerAddress: 127.0.0.1
    podSubnet: 10.244.0.0/16
    serviceSubnet: 10.96.0.0/16
    disableDefaultCNI: true
    kubeProxyMode: iptables
```

jpmcb committed:
unmanaged-cluster merging kind configs

- Prefers configs given via the ProviderConfiguration field and than
  attempts to merge any other default or flag given values into that
  kind configuration.

Signed-off-by: John McBride <jmcbride@vmware.com>

jpmcb commented:
cc @jorgemoralespou - would be awesome if you looked at this and checked if this is the UX you were hoping for! 👀

stmcginnis commented on cli/cmd/plugin/unmanaged-cluster/cluster/kind.go:
```diff
@@ -198,6 +202,69 @@ func kindConfigFromClusterConfig(c *config.UnmanagedClusterConfig) ([]byte, erro
+// into the kind config created via the provider config
+//
+// Provider configs take precidence.
+// So options given via `rawKindConfig` are the highest order configurations.
```
I wonder if it would be better to give command line arguments priority over rawKindConfig. That would allow someone to have a common base kind config that they always use that has the defaults they usually want, but then adjust it at deploy time by passing different arguments to the create command.

jpmcb replied:
> That would allow someone to have a common base kind config that they always use that has the defaults they usually want, but then adjust it at deploy time by passing different arguments to the create command.


That's a cool workflow. I like that idea. But I honestly have no preference. 

It would probably be good to define this in writing soon so we can have an expected _ethos_ users can refer to when creating configs / scripts, and we can have something to fall back on when these discussions come up.

Something like:
```
1. Command line arguments take precedence over all
2. Env variables take precedence over config files
3. Configuration files take precedence over default settings
4. Default values are used if nothing is provided
```

---

There is this comment which should probably go in the docs?
https://github.com/vmware-tanzu/community-edition/blob/690018ecd05cdbb8071e6e388463f6604a0abd24/cli/cmd/plugin/unmanaged-cluster/config/config.go#L168-L176

stmcginnis replied:
We do sort of have it documented here: https://tanzucommunityedition.io/docs/v0.11/ref-unmanaged-cluster/#custom-configuration

Not sure if someone would actually read that far, but since we do publicly state that, it would be good to stay consistent with that order of precedence.

jorgemoralespou replied:
I agree 100% with @stmcginnis as this is typical overriding of configuration in many programs I know and use.

stmcginnis reviewed (commented):
This looks good, will try to test it out soon.

jorgemoralespou commented:
Haven't tried it yet (will do shortly) but in the meantime, some quick questions:
- If I provide rawConfig like above with port mappings and IPs and then on the CLI I have `-p` flags with same ports only, what will happen?
- If I add via CLI or Config another port that is not in the rawKindConfig, will it be added to the resulting config?
- Will the rawKind networking be replaced to honour the UC-Config podCidr if both are provided?

jorgemoralespou commented:
I see some inconsistency when creating and deleting:

jorgemoralespou commented:
I did another test and got an error with compatibility cli:

```
./artifacts/darwin/amd64/cli/unmanaged-cluster/v0.10.0-dev.1/tanzu-unmanaged-cluster-darwin_amd64 create -f /tmp/ucconfig.yaml

📁 Created cluster directory
   Reading ProviderConfiguration from config file. Some provider specific flags and configs may be ignored.

🧲 Resolving and checking Tanzu Kubernetes release (TKr) compatibility file
   projects.registry.vmware.com/tce/compatibility
   Downloaded to: /Users/jomorales/.config/tanzu/tkg/unmanaged/compatibility/projects.registry.vmware.com_tce_compatibility_v4
failed parsing the TKr compatibility file. Error: could not find compatible CLI version in compatibility file
```

How can I test this?

jpmcb commented:
> I did another test and got an error with compatibility cli:

@jorgemoralespou - That is likely related to https://github.com/vmware-tanzu/community-edition/issues/3895

You can simply declare the TKR yourself with the `--tkr` flag and skip the compatibility check:
```
$  tanzu unmanaged-cluster create test --tkr projects.registry.vmware.com/tce/tkr:v0.17.0-dev-2
```

jpmcb commented:
> I'm now in a state where I can't create or delete this cluster. Will delete directory manually

That is a rough edge in the `delete` command that is being worked out in https://github.com/vmware-tanzu/community-edition/pull/3674 and is out of scope for this PR. 

Delete the directory yourself:

```
$ rm -rf ~/.config/tanzu/tkg/unmanaged/kind-test
```

And delete the cluster yourself via kind (if it was created and got to that step)

```
$ kind delete cluster --name kind-test
```

jpmcb commented:
> If I provide rawConfig like above with port mappings and IPs and then on the CLI I have -p flags with same ports only, what will happen?

> If I add via CLI or Config another port that is not in the rawKindConfig, will it be added to the resulting config?

`extraPortMappings` are really tricky since they are associated with nodes and attempting to merge different number of nodes could result in unexpected behavior. For example, if a user in the `rawKindConfig` has 3 nodes each with extra port mappings but than sets `--worker-node-count  5`, what would the expected behavior be?

My gut is that we should **_not_** attempt to merge nodes and port mappings but take them whole sale. If a user defines number of nodes via the config flags, we take those configurations whole sale and log to the user that we are ignoring the port mappings from the `rawKindconfig`.

_Maybe_ we could work around doing this if the number of number of nodes are the same and the number of port mappings given are the same.

But my question is this: @jorgemoralespou why would a user try to redefine the port mappings vai the CLI if they've already given them via the config file? Do you have a workflow example / user workflow you can give that shows how this might work?  

> Will the rawKind networking be replaced to honour the UC-Config podCidr if both are provided?

If both are provided, the CLI flags will take precedence over the `rawKindConfig`

jorgemoralespou commented:
> But my question is this: @jorgemoralespou why would a user try to redefine the port mappings vai the CLI if they've already given them via the config file? Do you have a workflow example / user workflow you can give that shows how this might work?

jpmcb committed:
Implements global config precedence

- Now, the global config via cli flags and env vars has precedence over
  the provider config file.
- Adds tests for merging of configs

Signed-off-by: John McBride <jmcbride@vmware.com>

jpmcb commented:
Ready for another round of reviews - the cli flags take precedence and we do a best effort 1 to 1 merging of nodes in the config. Give it a try and let me know if there are any sharp edges!

jorgemoralespou commented:
@jpmcb LGTM. My comments seem to have been addressed.

joshrosso commented:
working great on my end too. very nice :rocket:.

We may want to consider this for minikube too, at some point?

joshrosso commented:
working great on my end too. very nice :rocket:.

We may want to consider this for minikube too, at some point?
